apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt

```diff
@@ -0,0 +1,117 @@
+package com.apollographql.apollo3.network.ws
+
+import com.apollographql.apollo3.internal.ChannelWrapper
+import io.ktor.http.Headers
+import io.ktor.http.URLProtocol
+import io.ktor.http.Url
+import io.ktor.util.PlatformUtils
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.suspendCancellableCoroutine
+import okio.ByteString
+import org.khronos.webgl.Uint8Array
+import org.w3c.dom.MessageEvent
+import org.w3c.dom.WebSocket
+import org.w3c.dom.events.Event
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+actual class DefaultWebSocketEngine : WebSocketEngine {
+
+  override suspend fun open(
+      url: String,
+      headers: Map<String, String>,
+  ): WebSocketConnection = open(Url(url), headers)
+
+  private suspend fun open(url: Url, headers: Map<String, String>): WebSocketConnection {
+    val newUrl = url.copy(
+        protocol = when (url.protocol) {
+          URLProtocol.HTTPS -> URLProtocol.WSS
+          URLProtocol.HTTP -> URLProtocol.WS
+          URLProtocol.WS, URLProtocol.WSS -> url.protocol
+          /* URLProtocol.SOCKS */else -> throw UnsupportedOperationException("SOCKS is not a supported protocol")
+      })
+    val socket = createWebSocket(newUrl.toString(), Headers.build { headers.forEach { append(it.key, it.value) } }).awaitConnection()
+    val messageChannel = ChannelWrapper(Channel<String>(Channel.UNLIMITED))
+    socket.onmessage = { messageEvent: MessageEvent ->
+      val data = messageEvent.data
+      if (data != null) {
+        when (data) {
+          is String -> messageChannel.trySend(data)
+          else -> throw UnsupportedOperationException("${data::class.simpleName} is currently unsupported")
+        }
+      }
+    }
+    return object : WebSocketConnection {
+      override suspend fun receive(): String {
+        return messageChannel.receive()
+      }
+
+      override fun send(data: ByteString) {
+        socket.send(Uint8Array(data.toByteArray().toTypedArray()))
+      }
+
+      override fun send(string: String) {
+        socket.send(string)
+      }
+
+      override fun close() {
+        socket.close(CLOSE_NORMAL.toShort())
+        //close the message channel as well, since it is idempotent there is no harm
+        messageChannel.close()
+      }
+
+    }
+  }
+
+  /*
+   * The following applies for lines 71-114
+   * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+   * Some lines have been added/modified from the original source (https://github.com/ktorio/ktor/blob/f723638afd4d36024c390c5b79108b53ab513943/ktor-client/ktor-client-core/js/src/io/ktor/client/engine/js/JsClientEngine.kt#L62)
+   */
```

Comment:
LGTM. Maybe mention why the lines where modified?

```suggestion
  /*
   * The following applies for lines below
   * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
   * Some lines have been added/modified from the original source (https://github.com/ktorio/ktor/blob/f723638afd4d36024c390c5b79108b53ab513943/ktor-client/ktor-client-core/js/src/io/ktor/client/engine/js/JsClientEngine.kt#L62)
   * in order to fix an issue with subprotocols on Node (https://youtrack.jetbrains.com/issue/KTOR-4001)
   */
```